File: src/services/work.ts
import { WorkFormI } from '~/models/WorkForm'
import { CreateWorkInput, UpdateWorkInput, User } from '~/models/types'

export function UpdateWorkService(session: any, work: WorkFormI, UpdateWork: any): boolean {

  if (!session || !session.user) { return false }
  if (work.languages != undefined) {
    work.language = JSON.stringify(work.languages)
  }
  if (work.urls != undefined) {
    work.url = JSON.stringify(work.urls)
  }

  let updateWorkInput: UpdateWorkInput = {
    id: work.id,
    brief_story: work.brief_story,
    duration: work.duration,
    image_url: work.image_url,
    language: work.language,
    number_of_people: work.number_of_people,
    role: work.role,
    summary: work.summary,
    title: work.title,
    url: work.url,
  }
  UpdateWork({ variables: { input: updateWorkInput } })
  return true
}

export function CreateWorkService(session: any, work: WorkFormI, CreateWork: any): boolean {

  if (!session || !session.user) { return false }
  if (work.languages != undefined) {
    work.language = JSON.stringify(work.languages)
  }
  if (work.urls != undefined) {
    work.url = JSON.stringify(work.urls)
  }

  let createWorkInput: CreateWorkInput = {
    brief_story: work.brief_story,
    duration: work.duration,
    image_url: work.image_url,
    language: work.language,
    number_of_people: work.number_of_people,
    role: work.role,
    summary: work.summary,
    title: work.title,
    url: work.url,
    user_id: session.user.id,
  }
  CreateWork({
    variables: { input: createWorkInput },
  })
  return true
}

File: backend/domain/work.go
package domain

import "context"

type Work struct {
	ID             string  `json:"id"`
	Title          string  `json:"title"`
	Summary        *string `json:"summary"`
	ImageURL       *string `json:"imageUrl"`
	Duration       *string `json:"duration"`
	NumberOfPeople *int    `json:"numberOfPeople"`
	Language       *string `json:"language"`
	Role           *string `json:"role"`
	URL            *string `json:"url"`
	BriefStory     *string `json:"briefStory"`
	CreatedAt      string  `json:"createdAt"`
	UpdatedAt      string  `json:"updatedAt"`
	IsDelete       bool    `json:"isDelete"`
	UserID         string  `json:"userId"`
	SerialNumber   int     `json:"serialNumber" gorm:"autoIncrement"`
}

func (Work) IsNode()            {}
func (this Work) GetID() string { return this.ID }

// WorkUseCase represent the work's usecases.
type IWorkUseCase interface {
	GetByID(ctx context.Context, id string) (*Work, error)
	GetAll(ctx context.Context, sortBy SortBy, searchedAt string, num int, limit int) (*WorkPagination, error)
	Search(ctx context.Context, keyword string, sortBy SortBy, searchedAt string, num int, limit int) (*WorkPagination, error)
	Create(ctx context.Context, input CreateWorkInput) error
	Update(ctx context.Context, work *Work, input UpdateWorkInput) error
	Delete(ctx context.Context, ids []*string) error
}

// WorkRepository represent the work's repository contract.
type IWorkRepository interface {
	GetByID(ctx context.Context, id string) (*Work, error)
	GetTotalCount(ctx context.Context, keyword *string) (int64, error)
	GetAll(ctx context.Context, sortBy SortBy, searchedAt string, num int, limit int) ([]*Work, int64, error)
	GetByUserIDs(ids []string) ([]*Work, error)
	GetByKeyword(ctx context.Context, keyword string, sortBy SortBy, searchedAt string, num int, limit int) ([]*Work, int64, error)
	Create(ctx context.Context, input CreateWorkInput) error
	Update(ctx context.Context, work *Work, input UpdateWorkInput) error
	Delete(ctx context.Context, ids []*string) error
}
